### PR TITLE
add autocompletable header types

### PR DESCRIPTION
**Pull request by KhafraDev**
the header keys were taken from [MDN](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers) and the values for the content-type header were taken from [hono](https://github.com/honojs/hono/commit/ac069d1696536e6d8f43e9b360cbee389035e15f)

### CODE REVIEW

**Comment by KhafraDev** [comment]
The types only work for the object param of the constructor. It would be possible to add support for the other methods, but would require a generic which would interfere with the DOM types.

File: types/fetch.d.ts [hunk 1]
```diff
@@ -67,7 +67,7 @@ export interface SpecIterable<T> {
   [Symbol.iterator](): SpecIterator<T>;
 }
 
-export type HeadersInit = string[][] | Record<string, string | ReadonlyArray<string>> | Headers
+export type HeadersInit = [string, string][] | HeaderRecord | Headers
```
Comment (KhafraDev) [inline types/fetch.d.ts:70]: ReadonlyArray<string> is just not correct. If we were to follow the logic in the PR that added it, the type should be any.

Any array pair that doesn't contain exactly 2 strings throws an error too, so string[][] is not strict enough.

**Review by mcollina — APPROVED** [review]
lgtm

File: types/header.d.ts [hunk 2]
```diff
@@ -2,3 +2,155 @@
  * The header type declaration of `undici`.
  */
 export type IncomingHttpHeaders = Record<string, string | string[] | undefined>;
+
+type HeaderNames =
```
Comment (Uzlopak) [inline types/header.d.ts:6]: I think they should be lowercased?

File: types/header.d.ts [hunk 3]
```diff
@@ -2,3 +2,155 @@
...
+  | 'Upgrade-Insecure-Requests'
+  | 'User-Agent'
+  | 'Vary'
+  | 'Via'
+  | 'WWW-Authenticate'
+  | 'X-Content-Type-Options'
+  | 'X-Frame-Options'
+  | (string & {})
+
+type KnownHeaderValues = {
```
Comment (Uzlopak) [inline types/header.d.ts:98]: Why not a type ContentType ?

File: types/header.d.ts [hunk 4]
```diff
@@ -2,3 +2,155 @@
...
+    | 'application/zip'
+    | 'video/3gpp'
+    | 'video/3gpp2'
+    | 'model/gltf+json'
+    | 'model/gltf-binary'
+    | (string & {})
+}
+
+export type HeaderRecord = {
+  [K in HeaderNames | Lowercase<HeaderNames>]?: Lowercase<K> extends keyof KnownHeaderValues
```
Comment (Uzlopak) [inline types/header.d.ts:153]: Why do we lowercase them via a utiliy type?

File: types/header.d.ts [hunk 5]
```diff
@@ -2,3 +2,155 @@
...
+  | 'Transfer-Encoding'
+  | 'Upgrade'
+  | 'Upgrade-Insecure-Requests'
+  | 'User-Agent'
+  | 'Vary'
+  | 'Via'
+  | 'WWW-Authenticate'
+  | 'X-Content-Type-Options'
+  | 'X-Frame-Options'
+  | (string & {})
```
Comment (Uzlopak) [inline types/header.d.ts:96]: Maybe create a utility.d.ts if not exisiting already and create a Autocomplete type?

**Review by Uzlopak — CHANGES_REQUESTED** [review]
We should discuss this a little bit before merging.

File: types/header.d.ts (hunk 4 above)
Comment (Ethan-Arrowood) [reply types/header.d.ts:153]: I believe this expands to allow the user to specify either the uppercased (`Accept`) or lowercased (`accept`) option depending on their own personal preference. 

Per spec Headers can be specified either way and are normalized to be the same thing.

**Review by Ethan-Arrowood — APPROVED** [review]
lgtm

File: types/header.d.ts (hunk 2 above)
Comment (KhafraDev) [reply types/header.d.ts:6]: they are in the map, so it auto completes both the cased and lower cased values

File: types/header.d.ts (hunk 3 above)
Comment (KhafraDev) [reply types/header.d.ts:98]: I assumed that there would be more headers that we'd want to add types to

File: types/header.d.ts (hunk 4 above)
Comment (KhafraDev) [reply types/header.d.ts:153]: so we have intellisense for both 'Content-Type' and 'content-type'

**Pull request merged**